pageserver/src/tenant/blob_io.rs

```diff
@@ -199,27 +209,27 @@ impl<const BUFFERED: bool> BlobWriter<BUFFERED> {
                 let copied = self.write_into_buffer(&src_buf);
                 // We just verified above that src_buf fits into our internal buffer.
                 assert_eq!(copied, src_buf.len());
-                Slice::into_inner(src_buf)
+                src_buf
             } else {
                 let (src_buf, res) = self.write_all_unbuffered(src_buf, ctx).await;
                 if let Err(e) = res {
-                    return (src_buf, Err(e));
+                    return_!(src_buf, Err(e));
                 }
                 src_buf
             }
         } else {
-            Slice::into_inner(src_buf)
+            src_buf
         };
-        (src_buf, Ok(()))
+        return_!(src_buf, Ok(()));
     }
 
     /// Write a blob of data. Returns the offset that it was written to,
     /// which can be used to retrieve the data later.
-    pub async fn write_blob<B: BoundedBuf<Buf = Buf>, Buf: IoBuf + Send>(
+    pub async fn write_blob<Buf: IoBuf + Send>(
         &mut self,
-        srcbuf: B,
+        srcbuf: Slice<Buf>,
         ctx: &RequestContext,
-    ) -> (B::Buf, Result<u64, Error>) {
+    ) -> (Slice<Buf>, Result<u64, Error>) {
```

Comment:
No user of `blob_io` needs the `Slice`. Why change it?

Reply:
I guess there was a misunderstanding in [our 1:1 call](https://github.com/neondatabase/neon/pull/8225#discussion_r1661401819): I was strictly speaking about the `VirtualFile` API, you were probably thinking callers should be changed as well.

Reply:
Looking at the PR for why you're doing this one (https://github.com/neondatabase/neon/pull/8106) I think we should leave `write_blob`'s signature and `blob_io.rs'` `write_all` signature unchanged.

Reply:
> No user of blob_io needs the Slice. Why change it?

to expand, there is no way to obtain `B::Buf` from `B` without slicing (please tell me if there is such a way). If B is an empty buffer, slicing panics. With a `Slice`, at least I can return srcbuf here.

Reply:
>  (please tell me if there is such a way)

```rust
let foo: B = ...;
let underlying: B::Buf = foo.slice_full().into_inner();
```